### PR TITLE
Private DNS and Network Links for Onprem and Spoke

### PR DESCRIPTION
**Pull request by johnwildes**
I thought this would be a neat addition to the vwan-playground.  Private DNS links are not really necessary, but they add a nice "cloud native" way of doing lookups.  The zones themselves do auto registration for the VNET they are primarily assigned to and lookups for the secondary.  With the peerings you can lookup VM in either virtual network.  I've run this build several times and out of those tests, I encountered 2 instances where  the complete build failed on one of the zones, but I feel this was an ARM issue of timeout as the resources end up creating.

### CODE REVIEW

**Comment by StefanIvemo** [comment]
First off, thank you for contributing! I've notice lately that people actually use this and have plans to update the entire playground. When it was first created we didn't have loops and conditions in Bicep and it is a bit static. I plan to create a couple of different topologies that you can choose from when deploying. I plan to do this by making my modules a bit more reusable than they are today. 

Private DNS Zones sounds like a great addition, I'm all for it! I would like you to change the code a little bit though to support future changes to the playground. 

1.  `HubPrivateDNS` and `SpokePrivateDNS.bicep` are almost the same. Can you make them into one `PrivateDNS.bicep` and make the template a bit mor dynamic?

Maybe something like this could work (not tested):
```bicep
param zoneName string
param vnetLinks array

//create the dnszone resource
resource privateDnsZone 'Microsoft.Network/privateDnsZones@2020-06-01' = {
  name: zoneName
  location: 'global'
}

resource vnetLink 'Microsoft.Network/privateDnsZones/virtualNetworkLinks@2020-06-01' = [for link in vnetLinks: {
  name: '${zoneName}/${link.linkName}'
  location: 'global'
  properties: {
    registrationEnabled: link.registrationEnabled
    virtualNetwork: {
      id: link.vnetId
    }
  }
}]

output resourceId string = privateDnsZone.id
```

Then in main.bicep we can do something like this for each zone:
```bicep
var spokeDnsZoneVnetLinks = [
  {
    vnetId: spokevnet.outputs.id
    linkName: 'spoke-link'
    registrationEnabled: false
  }
  {
    vnetId: onpremvnet.outputs.id
    linkName: 'onprem-registration'
    registrationEnabled: true
  }
]

module spokednszone './PrivateDNS.bicep' = {
  name: 'spokednszonedeploy'
  scope: resourceGroup(spokerg.name)
  params:{
    zoneName: spokednszonename
    vnetLinks: spokeDnsZoneVnetLinks 
  }
}
```

2. Try using API version  `2020-06-01` for the DNS Zone creation. I believe that will be more stable and not cause the timeouts. 

Again, thank you so much for contributing. Really appreciate it! ❤

**Comment by johnwildes** [comment]
@StefanIvemo thank you for the feedback.  I'll get to work on it.  Your example makes it much clearer how it could be implemented.

**Comment by johnwildes** [comment]
@StefanIvemo wondering if you have any thoughts on this.  I'm getting this error when I run the deployment.  It creates both DNS zones, but fails to find priv.spoke.contoso.com.  
``` json
Deployment failed. Correlation ID: 5539c2fd-65de-45a8-b5cc-3d45f85c0de3. {
  "status": "Failed",
  "error": {
    "code": "ResourceDeploymentFailure",
    "message": "The resource operation completed with terminal provisioning state 'Failed'.",
    "details": [
      {
        "code": "DeploymentFailed",
        "message": "At least one resource deployment operation failed. Please list deployment operations for details. Please see https://aka.ms/DeployOperations 
for usage details.",
        "details": [
          {
            "code": "NotFound",
            "message": "{\r\n  \"error\": {\r\n    \"code\": \"ParentResourceNotFound\",\r\n    \"message\": \"Can not perform requested operation on nested resource. Parent resource 'priv.spoke.contoso.com' not found.\"\r\n  }\r\n}"
          },
          {
            "code": "NotFound",
            "message": "{\r\n  \"error\": {\r\n    \"code\": \"ParentResourceNotFound\",\r\n    \"message\": \"Can not perform requested operation on nested resource. Parent resource 'priv.spoke.contoso.com' not found.\"\r\n  }\r\n}"
          }
        ]
      }
    ]
  }
} 
```

Screenshots
image
image
image

image

It creates both zones but does not complete the links.  I've attempted to use 2020-06-01, and 2018-09-01 api same result.  

Thoughts?

File: playground/PrivateDNS.bicep [hunk 1]
```diff
@@ -0,0 +1,21 @@
+param zoneName string
+param vnetLinks array
+
+//create the dnszone resource
+resource privateDnsZone 'Microsoft.Network/privateDnsZones@2018-09-01' = {
+  name: zoneName
+  location: 'global'
+}
+
```
Comment (StefanIvemo) [inline playground/PrivateDNS.bicep:9]: Try to add `@batchSize(1)` here to make sure that the VNet links are deploy in serial mode and not both at the same time.

File: playground/main.bicep [hunk 2]
```diff
@@ -275,6 +304,23 @@ module spokevnet './SpokeVNet.bicep' = {
...
+  scope: resourceGroup(spokerg.name)
+  params:{
+    zoneName: spokednszonename
+    vnetLinks: spokeDnsZoneVnetLinks 
+  }
+}
+
+module onpremdnszone './PrivateDNS.bicep' = {
+  name: 'onpremdnszonedeploy'
+  scope: resourceGroup(onpremrg.name)
```
Comment (StefanIvemo) [inline playground/main.bicep]: You can also add a `dependsOn` here to make sure that the DNS Zones are deployed separately to make sure that there are no conflicting actions on the VNets.
```bicep
  dependsOn:[
    spokednszone 
  ]
```

**Comment by StefanIvemo** [comment]
@johnwildes my first thought without testing is that it tries to configure all VNet links at the same time and that the VNet is busy updating, hence the conflict. I would try to add `@batchsize()`  to the loop inside the `PrivateDNS.bicep` and also add a `dependsOn` for the DNS Zones (see comments).

File: playground/PrivateDNS.bicep (hunk 1 above)
Comment (johnwildes) [reply playground/PrivateDNS.bicep:9]: Thanks, will try that.  Where do you find the documentation for those parameter guidelines, I see an example of this in the BICEP doc where they limit characters for things like a storage account name.  I'm wondering what else you can do, given the example you've listed above.

File: playground/main.bicep (hunk 2 above)
Comment (johnwildes) [reply playground/main.bicep]: This I did as well.  Working on testing now.

File: playground/PrivateDNS.bicep (hunk 1 above)
Comment (johnwildes) [reply playground/PrivateDNS.bicep:9]: I'm getting an error trying to use `@batchsize(1)` - `cannot be used as a param decorator`

Comment (StefanIvemo) [reply playground/PrivateDNS.bicep:9]: @batchsize(1) is a decorator that should be added to a resource declaration using loops. You can find more information about it here: https://github.com/Azure/bicep/blob/main/docs/spec/loops.md#batch-size-decorator

Decorators that can be used for parameters can be found here.

**Comment by StefanIvemo** [comment]
Is it working without conflict now? Is it ready for testing?

**Comment by johnwildes** [comment]
It is working without conflict now.  I ran it twice, no errors.

**Comment by StefanIvemo** [comment]
Nice! I will test this tomorrow and verify that it works as expected. Great job, and thanks again for contributing!

**Comment by StefanIvemo** [comment]
Looks good! Thanks a lot!

**Pull request merged**